### PR TITLE
Write out some set_ overloads instead of relying on code binding generation.

### PR DESCRIPTION
**Pull request by gchanan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28597 Kill _th_zero binding, just use a simple native function instead.
* #28596 Kill THTensor_(match), which isn't used.
* **#28595 Write out some set_ overloads instead of relying on code binding generation.**

Differential Revision: [D18116720](https://our.internmc.facebook.com/intern/diff/D18116720)

### CODE REVIEW

**Comment by gchanan** [comment]
Failures look plausibly caused by this, although I haven't been able to reproduce them locally.

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -34,6 +34,15 @@ Tensor _shape_as_tensor(const Tensor& self) {
   return at::tensor(self.sizes(), options);
 }
 
+Tensor& set_(Tensor& result) {
+  Storage storage(result.dtype(), 0, c10::GetAllocator(result.device().type()), true);
```
Comment (albanD) [inline aten/src/ATen/native/TensorShape.cpp]: Isn't the storage supposed to be a nullptr from the previous signature in case it is not given by the user?

Comment (gchanan) [reply aten/src/ATen/native/TensorShape.cpp]: yes, but we can't don't actually support passing nullptrs in this way through the ATen->TH binding layer.

So what this PR attempts to do instead is to create a Storage in the same way the TH code does if you pass in a nullptr (you can't have nullptr Storage, so TH does this logic already).

Comment (albanD) [reply aten/src/ATen/native/TensorShape.cpp]: Right, the fix looks like the right thing to do :)

**Review by albanD — APPROVED** [review]

**Pull request closed without merging**